Mesh meta data passing

miaoyinb opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
Mesh modifiers do not retain input mesh's metadata, which cases issues.

## Design
Establish a mesh metadata alias system to allow retainment of mesh metadata during modifications.

## Impact
Facilitate use of mesh metadata for consistent mesh generation.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

miaoyinb commented:
@GiudGiud
This is the metadata retaining one. Not as developed as the assignment PR (no documentation/testing yet). But you can see how the alias system I am proposing works. Thank you.

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -20,15 +20,58 @@ MeshGenerator::validParams()
+      "Names of input mesh's metadata that are selected to retain.");
+  params.addParam<bool>("retain_all_input_mesh_metadata",
+                        false,
+                        "Whether to retain all the mesh metadata of the input mesh.");
```
you ll want a parameter group

GiudGiud commented on modules/reactor/include/meshgenerators/MeshPropertyPassing.h:
```diff
@@ -0,0 +1,31 @@
+#include "PolygonMeshGeneratorBase.h"
+#include "MooseEnum.h"
+
+/**
+ * This MeshPropertyPassing object
+ */
+class MeshPropertyPassing : public PolygonMeshGeneratorBase
+{
```
what is this object for?

miaoyinb replied:
Please disregard this. It is just for my own testing purpose. Will not be included in the formal PR

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -20,15 +20,58 @@ MeshGenerator::validParams()
+  : MooseObject(parameters),
+    MeshMetaDataInterface(this),
+    _mesh(_app.actionWarehouse().mesh()),
+    _selected_mesh_metadata_to_retain(
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  if (isParamValid("input"))
+  {
+    const MeshGeneratorName input_name(getParam<MeshGeneratorName>("input"));
+    if (_retain_all_input_mesh_metadata)
+    {
+      if (!_selected_mesh_metadata_to_retain.empty())
+        paramError(
+            "selected_mesh_metadata_to_retain",
+            "This parameter should not be provided if retain_all_input_mesh_metadata is set true.");
+      const auto mesh_metadata_names = findMeshMetaData(input_name);
+      for (const auto & mmd_name : mesh_metadata_names)
+        AddMeshMetaDataAlias(input_name, mmd_name, name(), mmd_name);
+    }
+    else
+      for (const auto & mmd_name : _selected_mesh_metadata_to_retain)
+      {
+        if (!hasMeshProperty(mmd_name, input_name))
+          paramError("selected_mesh_metadata_to_retain",
+                     "The specified mesh metadata to retain does not exist in the input mesh.");
+        AddMeshMetaDataAlias(input_name, mmd_name, name(), mmd_name);
+      }
+  }
```
we should move some or most of this code inside the interface.
It s cleaner to have small minimal base classes

GiudGiud commented on framework/include/interfaces/MeshMetaDataInterface.h:
```diff
@@ -59,6 +59,40 @@ class MeshMetaDataInterface
+   * @param new_prefix the new prefix (mesh generator name) of the mesh metadata alias
+   * @param new_name the new mesh metadata name of the mesh metadata alias
+   */
+  void AddMeshMetaDataAlias(std::string original_prefix,
```
```suggestion
  void addMeshMetaDataAlias(std::string original_prefix,
```

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -20,15 +20,58 @@ MeshGenerator::validParams()
+        paramError(
+            "selected_mesh_metadata_to_retain",
+            "This parameter should not be provided if retain_all_input_mesh_metadata is set true.");
+      const auto mesh_metadata_names = findMeshMetaData(input_name);
```
so this constructor is run before the derived class'
so do we have the metadata set at this time?

miaoyinb replied:
You are right. This would only work for metadata declared in initializer list. This is the case for many reactor module MGs. But I need to think about a more generalized approach.

miaoyinb replied:
Moved to `MeshGenerator::generateInternal()`

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -115,6 +139,30 @@ MeshGenerator::buildDistributedMesh(unsigned int dim)
 std::unique_ptr<MeshBase>
 MeshGenerator::generateInternal()
 {
+  // Retain mesh metadata from input if applicable
+  if (isParamValid("input"))
+  {
+    const MeshGeneratorName input_name(getParam<MeshGeneratorName>("input"));
+    if (_retain_all_input_mesh_metadata)
+    {
+      if (!_selected_mesh_metadata_to_retain.empty())
+        paramError(
+            "selected_mesh_metadata_to_retain",
+            "This parameter should not be provided if retain_all_input_mesh_metadata is set true.");
+      const auto mesh_metadata_names = findMeshMetaData(input_name);
+      for (const auto & mmd_name : mesh_metadata_names)
+        addMeshMetaDataAlias(input_name, mmd_name, name(), mmd_name);
+    }
+    else
+      for (const auto & mmd_name : _selected_mesh_metadata_to_retain)
+      {
+        if (!hasMeshProperty(mmd_name, input_name))
+          paramError("selected_mesh_metadata_to_retain",
+                     "The specified mesh metadata to retain does not exist in the input mesh.");
+        addMeshMetaDataAlias(input_name, mmd_name, name(), mmd_name);
+      }
+  }
```
I would still move that code inside the interface. And call it from here

miaoyinb replied:
will do

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -20,15 +20,39 @@ MeshGenerator::validParams()
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  if (!isParamValid("input"))
+  {
+    if (_retain_all_input_mesh_metadata)
+      paramError("retain_all_input_mesh_metadata",
+                 "In the absence of an input mesh, this parameter must not be true.");
+    if (!_selected_mesh_metadata_to_retain.empty())
+      paramError("selected_mesh_metadata_to_retain",
+                 "In the absence of an input mesh, this parameter must be empty.");
+  }
```
I think this can be done from the inerface constructor, if you pass the parameters

miaoyinb commented:
Moved to `MeshMetaDataInterface`

miaoyinb commented:
@GiudGiud As you pinned me in another PR, just a reminder that this is the local metadata forwarding PR that will be submitted once you think it is ready. (in case it's confusing, the one I just submitted is just for adding new metadata).

GiudGiud commented:
I think this is relevant to what @shikhar413 is doing

shikhar413 commented:
> I think this is relevant to what @shikhar413 is doing

Am aware of this effort. Haven't looked into using this in detail yet but it should remove the need for some of the logic in my other PR

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -20,15 +20,34 @@ MeshGenerator::validParams()
                         "Whether or not to show mesh info after generating the mesh "
                         "(bounding box, element types, sidesets, nodesets, subdomains, etc)");
+  params.addParam<std::vector<std::string>>(
+      "selected_mesh_metadata_to_retain",
+      std::vector<std::string>(),
+      "Names of input mesh's metadata that are selected to retain.");
+  params.addParam<bool>("retain_all_input_mesh_metadata",
+                        false,
+                        "Whether to retain all the mesh metadata of the input mesh.");
+
+  params.addParamNamesToGroup("selected_mesh_metadata_to_retain retain_all_input_mesh_metadata",
+                              "Mesh Metadata Retainment");
 
   params.registerBase("MeshGenerator");
 
   return params;
 }
 
 MeshGenerator::MeshGenerator(const InputParameters & parameters)
-  : MooseObject(parameters), MeshMetaDataInterface(this), _mesh(_app.actionWarehouse().mesh())
+  : MooseObject(parameters),
+    MeshMetaDataInterface(this),
+    _mesh(_app.actionWarehouse().mesh()),
+    _selected_mesh_metadata_to_retain(
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  checkMeshMetadataRetainingSetting(this,
+                                    isParamValid("input"),
+                                    _retain_all_input_mesh_metadata,
+                                    _selected_mesh_metadata_to_retain);
```
could we put all the parameters in the interface?

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -111,6 +111,12 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
   /// References to the mesh and displaced mesh (currently in the ActionWarehouse)
   std::shared_ptr<MooseMesh> & _mesh;
 
+  /// List of selected mesh metadata from the input mesh that need to be retained
+  const std::vector<std::string> _selected_mesh_metadata_to_retain;
+
+  /// Whether to retain all the mesh metadata from the input mesh
+  const bool _retain_all_input_mesh_metadata;
```
same cant it be in the interface?

GiudGiud reviewed (commented):
What's the motivation for creating the "Alias" system? Why cant you reuse the existing map?

miaoyinb commented:
> What's the motivation for creating the "Alias" system? Why cant you reuse the existing map?

Let me look into this to see if things can be done directly in `RestartableDataMap`.

I had an impression that it would mess up something but I could be wrong. It has been a while.

miaoyinb commented:
> What's the motivation for creating the "Alias" system? Why cant you reuse the existing map?

`RestartableDataMap` is `std::unordered_map<std::string, RestartableDataValuePair<std::unique_ptr<RestartableDataValue>, bool>>`. Because of the unique pointer, I do not think we can make reuse it. A separate alias system would make more sense to me. Let me know your idea. Thanks.

GiudGiud commented:
Alias system seems problematic to me. You end up having to worry about whether they are synchronized. The routines are somewhat duplicated
```
findMeshMetaDataAlias / findMeshMetaData
hasMeshMetaDataAlias / hasMeshMetaData
```

Unique pointer will forbid you from holding it as a class attribute in the class. However with @shikhar413 's work we are going towards not doing that when possible. We should be able to query it and modify it, not keeping ownership local to the class ?

miaoyinb commented:
> 



> Alias system seems problematic to me. You end up having to worry about whether they are synchronized. The routines are somewhat duplicated
> 
> ```
> findMeshMetaDataAlias / findMeshMetaData
> hasMeshMetaDataAlias / hasMeshMetaData
> ```
> 
> Unique pointer will forbid you from holding it as a class attribute in the class. However with @shikhar413 's work we are going towards not doing that when possible. We should be able to query it and modify it, not keeping ownership local to the class ?

I can make integrate the Alias related stuff into findMeshMetaData and hasMeshMetaData so it is less confusing.

Unique pointer cannot be copied so it is challenging to have two elements in `RestartableDataMap` sharing the same pointer but different names. Any suggestions?

GiudGiud commented:
I think you have to copy the metadata anyway, you cant share the pointer.

If you have a mesh from MG1 then you have two mesh generators, MG2 and MG3 that both use the MG1 mesh. 
If MG2 modifies the metadata and the pointer is shared, then that data is modified for the input of MG3 as well.

miaoyinb commented:
> I think you have to copy the metadata anyway, you cant share the pointer.
> 
> If you have a mesh from MG1 then you have two mesh generators, MG2 and MG3 that both use the MG1 mesh. If MG2 modifies the metadata and the pointer is shared, then that data is modified for the input of MG3 as well.

I agree if a metadata needs to be modified, it has to be copied. But this is not the focus of this PR. In this PR, I want to enable retaining or forwarding of metadata (so no modification).

Copying a metadata data is not trivial, we have to specify the data type. In my opinion, only two types of objects should know the data type of a metadata: its generator and its consumer (by consumer, I also covers those objects that need to modify a metadata). For any other intermediate MGs (e.g., add a boundary or modify a block name...), they do not need to (and are probably not able to) know the metadata type of their input mesh. Thus, copying a metadata becomes challenging. This is why I developed this "Alias" based approach so that the intermediate MGs can retain/forward the input mesh's metadata to the real consumer without know the metadata type.

GiudGiud commented:
The type can be obtained from the map. @shikhar413's PR should have done that somewhere now

Copy is safer. The distinction between the two is generally too difficult for users.
I also think that we are in a situation where we could use the copy. Reactors often involve lattices at one or two levels and the  metadata of an element of this lattice could be modified in one location and not modified in another

shikhar413 commented:
@miaoyinb for reference, in my other PR I obtain the original data type of the metadata like follows:
`_app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0).type()`. This returns the datatype as a string

miaoyinb commented:
> @miaoyinb for reference, in my other PR I obtain the original data type of the metadata like follows: `_app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0).type()`. This returns the datatype as a string

Thanks, let me take a look into this.

miaoyinb commented:
@GiudGiud @shikhar413 
I took a look at this. The `type()` function does give `std::string` type information of the metadata type. However, this information cannot be used in the templates such as `declareMeshProperty()` because templates are compile time constructs. So if we want to copy the metadata, a different approach will be needed. Any suggestion would be appreciated.

GiudGiud commented:
You could do something like:
```
str::string type = _app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0).type();
if (type == string (the name) )
{
  T & data = _app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0)
  declareMeshProperty<T>(data); 
}
```
for each type. It's not super clean. 

Another idea would be to rework the base map with std::any or variants. 
This might be more work and less efficient

miaoyinb commented:
> You could do something like:
> 
> ```
> str::string type = _app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0).type();
> if (type == string (the name) )
> {
>   T & data = _app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0)
>   declareMeshProperty<T>(data); 
> }
> ```
> 
> for each type. It's not super clean.
> 
> Another idea would be to rework the base map with std::any or variants. This might be more work and less efficient

Thanks, @GiudGiud. It seems the first approach is more efficient although we need to traverse all the types that have been used as Mesh Metadata.

Maybe we should add a test that check if any new types are used so that developers can add new types to this part.

GiudGiud commented:
We can add an error when the type is not recognized in the "switch" condition. 
It still is not super satisfying, just like defining all the metadata type by type is not very satisfying. 

As it s an existing "defect" I wont object to doing it that way. 
But we would definitely welcome a long term solution.

miaoyinb committed:
add needed methods

miaoyinb committed:
working

miaoyinb committed:
test and doc

miaoyinb commented:
@GiudGiud 

Updated this PR with the new "copying" approach you suggested. Still need some formalization but please take a look to see if the implementation makes sense.

I added a check in `declareMeshProperty` to ensure that all the metadata types are covered. It may trigger more errors in other apps that declare their own metadata but these can be fixed once we do the civet running. (Maybe this check can be set to only run in DEBUG mode.)

Thanks.

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -213,6 +327,212 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
+{
+  switch (type_id)
+  {
+    case 0:
```
this is prone to error. Can we switch on the name instead?

miaoyinb replied:
We cannot switch on strings. Maybe use enum?

miaoyinb replied:
Or I can make a function for this.

GiudGiud replied:
enum seems fine too. Just not mapping manually from a string to an int

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -68,6 +68,92 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
+      "std::vector<std::vector<unsigned short>>",
+      "std::vector<std::vector<std::string>>",
+      "std::map<short, libMesh::VectorValue<double>, std::less<short>, "
+      "std::allocator<std::pair<short const, libMesh::VectorValue<double> > > >"};
```
looping through the map, can we build this list automatically?

miaoyinb replied:
That's a good idea.

miaoyinb replied:
@GiudGiud I managed to use a string-enum map to do the work instead of a manual mapping process.

I thought about your question here. The metadata map is created in MG constructors. So it only contains metadata that are used in the specific input file. However, we want a list here that covers everything and it will be used during compilation (as we are still "manually" giving the typenames).

GiudGiud reviewed (commented):
dont worry about other apps and mesh metadata
Only reactor physics is using this right now

miaoyinb commented:
> dont worry about other apps and mesh metadata Only reactor physics is using this right now

Some BISON classes do use mesh metadata. (I confess that I made them :sweat_smile:).

miaoyinb committed:
update 1

miaoyinb committed:
update 2

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -213,6 +367,252 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
+inline bool
+MeshGenerator::typeDoubleCheck(T temp_value, const std::string type_name)
+{
+  (void)temp_value;
```
what is this for

miaoyinb replied:
Just to get the type through template. `{void)` is to "use" it so that the compiler does not complain.

GiudGiud replied:
Can you do MeshGenerator::typeDoubleCheck(T /* temp_value */, 

?

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -213,6 +367,252 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
 
+template <typename T>
+inline bool
+MeshGenerator::typeDoubleCheck(T temp_value, const std::string type_name)
```
why "double" check?

miaoyinb replied:
Although we are using a <string, Enum> map, it is still kinda of "manual". This double check is just to compare the declared type with the original type automatically to ensure no mistakes are made.

miaoyinb replied:
It is not required. I am okay not implement this if you do not like it.

GiudGiud replied:
I m fine with it, it's just the name that seems odd? It's just a check

miaoyinb replied:
`meshMetadataTypeCheck` seems to be a better name

GiudGiud commented on framework/include/interfaces/MeshMetaDataInterface.h:
```diff
@@ -59,6 +59,26 @@ class MeshMetaDataInterface
+   * need to be retained
+   */
+  void checkMeshMetadataForwardingSetting(
+      const MooseObject * moose_object,
```
isnt there a smaller class for this? like MeshGenerator?

GiudGiud commented on framework/include/interfaces/MeshMetaDataInterface.h:
```diff
@@ -59,6 +59,26 @@ class MeshMetaDataInterface
+   */
+  void checkMeshMetadataForwardingSetting(
+      const MooseObject * moose_object,
+      const bool has_input,
```
you can get that from the parameters of the object?

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -68,6 +68,132 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
   template <typename T>
   T & declareMeshProperty(const std::string & data_name, const T & init_value);
 
+  /// eunm class for type names
```
```suggestion
  /// enum class for all types that can be used as mesh metadata
```

GiudGiud commented on framework/include/meshgenerators/MeshGenerator.h:
```diff
@@ -68,6 +68,132 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
+    typeMapBoundaryIDRealVectorValue
+  };
+  /// This collects all the type names that are declared as Mesh Metadata
+  const std::map<std::string, type_names_enum> _type_name_map = {
```
you should be able to use a MooseEnum instead

GiudGiud replied:
umm maybe not.
I think that might force you to use integers again

GiudGiud commented on framework/src/interfaces/MeshMetaDataInterface.C:
```diff
@@ -34,3 +34,42 @@ MeshMetaDataInterface::hasMeshProperty(const std::string & data_name,
   return _meta_data_app.hasRestartableMetaData(full_name, MooseApp::MESH_META_DATA);
 }
+
+void
+MeshMetaDataInterface::checkMeshMetadataForwardingSetting(
+    const MooseObject * moose_object,
+    const bool has_input,
+    const bool retain_all_input_mesh_metadata,
+    const std::vector<std::string> selected_mesh_metadata_to_retain)
+{
+  if (!has_input)
+  {
+    if (retain_all_input_mesh_metadata)
+      moose_object->paramError("retain_all_input_mesh_metadata",
+                               "In the absence of an input mesh, this parameter must not be true.");
+    if (!selected_mesh_metadata_to_retain.empty())
+      moose_object->paramError("selected_mesh_metadata_to_retain",
+                               "In the absence of an input mesh, this parameter must be empty.");
+  }
+  else if (retain_all_input_mesh_metadata && !selected_mesh_metadata_to_retain.empty())
+    moose_object->paramError(
+        "selected_mesh_metadata_to_retain",
+        "This parameter should not be provided if retain_all_input_mesh_metadata is set true.");
+}
+
+std::vector<std::string>
+MeshMetaDataInterface::identifyMeshMetaData(const MeshGeneratorName prefix) const
+{
+  std::vector<std::string> res;
+  const std::string full_prefix = std::string(SYSTEM) + "/" + prefix + "/";
+  const RestartableDataMap & meta_data =
+      _meta_data_app.getRestartableDataMap(MooseApp::MESH_META_DATA);
+  for (const auto & meta_data_item : meta_data)
+  {
+    if (meta_data_item.first.size() > full_prefix.size())
+      if (meta_data_item.first.compare(0, full_prefix.size(), full_prefix) == 0)
+        res.push_back(meta_data_item.first.substr(
+            full_prefix.size(), meta_data_item.first.size() - full_prefix.size()));
+  }
+  return res;
```
so as was discussed with Cody on Shikhar s PR, this MeshMetaDataInterface is only for consumers of metadata, not producers.

so let s moved the code related to forwarding back to MG.C

GiudGiud replied:
maybe identify is more of a consumer thing though
what do you think

miaoyinb replied:
I agree with you that that `identifyMeshMetaData` can stay in the interface.

GiudGiud commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -20,15 +20,38 @@ MeshGenerator::validParams()
+                                     isParamValid("input"),
+                                     _retain_all_input_mesh_metadata,
+                                     _selected_mesh_metadata_to_retain);
+  if (_retain_all_input_mesh_metadata)
+    declareAllForwardedMeshMetadata(
+        getParam<MeshGeneratorName>("input"), _forwarded_metadata_names, _forwarded_metadata_types);
+  else if (!_selected_mesh_metadata_to_retain.empty())
+    declareSelectedForwardedMeshMetadata(getParam<MeshGeneratorName>("input"),
+                                         _selected_mesh_metadata_to_retain,
+                                         _forwarded_metadata_names,
+                                         _forwarded_metadata_types);
```
this still runs before most metadata is declared by the MG.

The MG base constructor is called before meta data declaration

miaoyinb replied:
The `input` MG constructor must be run before the MG constructor that needs to copy `input` mesh's metadata. As metadata can only be declared in constructor, we should be safe here.

miaoyinb committed:
code improvement and move some class back to MG

miaoyinb commented:
@GiudGiud Let me know if any other revision is needed before moving it to the official PR stage.

GiudGiud commented:
Logan worked up a design that will fit both declare and setting stage without the need for these enums and manual maps. He ll be in touch once it's clear if it will work or not. To be fair what you have now is pretty good, but we could condense this down quite a bit if Logan's solution works

GiudGiud commented:
@loganharbour

loganharbour commented:
Yeah, I've got a good idea for this. A WIP, but in progress nonetheless. Gets rid of the maps.
